### PR TITLE
refactor: update `libanki/Card.kt` partially 

### PR DESCRIPTION
**Pull request by BrayanDSO**
## Purpose / Description

Be up to date with libanki's `Cards.py`

## Approach

Didn't update all of it because it is necessary to do some functional changes about how the CardBrowser deals with `OpChanges`, and the browser is going through a major refactor already, so it wasn't worthwhile

## How Has This Been Tested?

Trusting CI

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.kt [hunk 1]
```diff
@@ -1317,6 +1331,19 @@ class ContentProviderTest : InstrumentedTest() {
             )
         }
     }
+
+    fun addNonClozeModel(name: String, fields: Array<String>, qfmt: String?, afmt: String?): String {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.kt:1335]: Shouldn't you also delete the current `addNonClozeModel` in the same commit ?

File: AnkiDroid/src/main/java/com/ichi2/libanki/Deck.kt [hunk 2]
```diff
@@ -43,7 +43,7 @@ class Deck : JSONObject {
     private constructor() : super()
 
     val isFiltered: Boolean
-        get() = getInt("dyn") == Consts.DECK_DYN
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Deck.kt:46]: Calling this commit "refactor" seems wrong to me.

Add-ons can have many strange effect. And I can imagine that for some reason someone thought wise to store some value, different from 0 and 1, in `dyn`, and checked it still works in anki. In which case, you may be ensuring the expected behavior on ankidroid. Or not, maybe on the contrary, they wanted to see a different behaviour, hard to tell.

I'll admit that this is a correct refactoring when the value is either 0 or 1. 
And while I believe code is less readable, this is closer to actually upstream implementation, so I won't object to get this merged. I just want a note about the behaviour change

**Review by Arthur-Milchior — COMMENTED** [review]
I started reviewing it before it was moved to draft. As it's now draft, I stop here

File: AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.kt (hunk 1 above)
Comment (BrayanDSO) [reply AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.kt:1335]: it is used somewhere else.

File: AnkiDroid/src/main/java/com/ichi2/libanki/Deck.kt (hunk 2 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Deck.kt:46]: Upstream is:

```python
    def is_filtered(self, did: DeckId | str) -> bool:
        return bool(self.get(did)["dyn"])
```

https://github.com/ankitects/anki/blob/74a603e61bb4dfbb22c443635ffef6e42a3ae3c5/pylib/anki/decks.py#L528-L529

As per https://docs.python.org/3/library/stdtypes.html#truth, this code matches

>  Here are most of the built-in objects considered false:
> - ... 
>  - zero of any numeric type: 0, 0.0, 0j, Decimal(0), Fraction(0, 1)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt [hunk 3]
```diff
@@ -360,7 +360,9 @@ open class Card : Cloneable {
     }
 
     fun userFlag(): Int {
-        return intToFlag(flags)
+        // equivalent to `mFlags % 8`. Used this way to copy Anki.
+        // setting all bits to 0, except the three first one.
+        return flags and 7
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:340]: Kotlin supports `0b111`, which would both be more accurate, and match upstream

File: AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt [hunk 4]
```diff
@@ -93,49 +93,22 @@ open class Card : Cloneable {
...
-    constructor(col: Collection, id: Long?) {
-        timerStarted = 0L
-        renderOutput = null
-        note = null
+    constructor(col: Collection, id: Long? = null) {
         if (id != null) {
             this.id = id
             load(col)
         } else {
-            // ephemeral card
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:140]: Just to confirm: TemplatePreviewerViewModel still works after this change?

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/RescheduleDialog.kt [hunk 5]
```diff
@@ -73,13 +73,13 @@ class RescheduleDialog : IntegerDialog() {
...
         private fun getContentString(resources: Resources, currentCard: Card): String? {
-            if (currentCard.isNew) {
+            if (currentCard.type == Consts.CARD_TYPE_NEW) {
                 return resources.getString(R.string.reschedule_card_dialog_new_card_warning)
             }
 
             // #5595 - Help a user reschedule cards by showing them the current interval.
             // DEFECT: We should be able to calculate this for all card types - not yet performed for non-review or dynamic cards
-            if (!currentCard.isReview) {
+            if (!(currentCard.type == Consts.CARD_TYPE_REV && currentCard.queue == Consts.QUEUE_TYPE_REV)) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/RescheduleDialog.kt:82]: I don't like it, but we're removing it soon

File: AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt [hunk 6]
```diff
@@ -269,8 +269,8 @@ open class Card : Cloneable {
     }
 
     @VisibleForTesting
-    fun setReps(reps: Int): Int {
-        return reps.also { this.reps = it }
+    fun setReps(reps: Int) {
+        this.reps = reps
     }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:274]: Can we use `var reps`

File: AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt (hunk 4 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:140]: Yeah. The tests in `TemplatePreviewerViewModelTest` depend on that

**Review by david-allison — APPROVED** [review]
As confirmation: squash merge

**Review by krmanik — APPROVED** [review]
I have done review commit by commit and it looks good to me. Thanks

**Pull request merged**